### PR TITLE
[haku] Squared rel-L2 aux loss for hard-sample focusing

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Adding a squared rel-L2 auxiliary loss term penalizes large per-sample errors more than the current MSE/rel-L2 loss, which should shrink the long tail of high-shear cases that drive `wall_shear_y_rel_l2_pct` and `wall_shear_z_rel_l2_pct` upward.**

The current Wave 1 winner uses an MSE-style loss in normalized space + (in some PRs) a raw rel-L2 aux loss with weight 0.05–0.3. The validation metric is `rel_l2_pct = ||pred - target|| / ||target||` averaged over samples. **MSE penalizes squared error, but rel-L2 already includes a per-sample normalization that softens the gradient signal at hard samples.**

Squared rel-L2:
```
sq_rel_l2(pred, target) = (||pred - target||^2 / (||target||^2 + eps))
```
This is the actual rel-L2 ratio squared, so:
- It maps directly onto a transformation of the eval metric.
- It places **strictly more weight on the worst-predicted samples** (long-tail high-shear cars) than rel-L2 does.
- Combined with the existing MSE base loss, it acts like a focal-loss equivalent for regression — easy samples see normal gradient, hard samples see amplified gradient.

This is orthogonal to:
- norman PR #180 (raw rel-L2 aux loss): linear rel-L2, doesn't focus on hard samples
- senku PR #145 (mse + raw rel-L2 aux 0.05): linear rel-L2 again
- Loss multipliers (edward PR #236): rebalances per-CHANNEL but not per-SAMPLE difficulty

## Instructions

**Implementation in `train.py`**:

1. Add CLI flag `--sq-rel-l2-aux-weight` (default 0.0, disabled).
2. In the loss computation, after the existing MSE/rel-L2 base loss, add:
```python
def squared_rel_l2(pred, target, eps=1e-6):
    # pred, target: (B, N, C) — sum over N and C, mean over B
    diff_sq = ((pred - target) ** 2).sum(dim=(-2, -1))
    target_sq = (target ** 2).sum(dim=(-2, -1)) + eps
    return (diff_sq / target_sq).mean()

if args.sq_rel_l2_aux_weight > 0:
    sq_rel_l2_loss = (
        squared_rel_l2(pred_surf_p, tgt_surf_p)
        + squared_rel_l2(pred_ws, tgt_ws)
        + squared_rel_l2(pred_vol_p, tgt_vol_p)
    ) / 3.0
    total_loss = base_loss + args.sq_rel_l2_aux_weight * sq_rel_l2_loss
    log("train/sq_rel_l2_loss", sq_rel_l2_loss)
```

3. Compute the squared-rel-L2 aux loss in **normalized output space** (same space as the base loss). Do NOT mix normalized and raw spaces.

**Run three trials** in W&B group `bengio-wave3-haku-sq-rel-l2`:

- **Trial A**: `--sq-rel-l2-aux-weight 0.1`
- **Trial B**: `--sq-rel-l2-aux-weight 0.3` ← **primary hypothesis**
- **Trial C**: `--sq-rel-l2-aux-weight 1.0`

All three on the bengio Wave 1 winning recipe (4L/256d + ContinuousSincosEmbed + no-EMA + cosine T_max=30):

```bash
cd target/ && python train.py \
  --model-num-layers 4 \
  --model-hidden-dim 256 \
  --no-use-ema \
  --lr-cosine-t-max 30 \
  --no-compile-model \
  --sq-rel-l2-aux-weight <0.1|0.3|1.0> \
  --nproc_per_node 4 \
  --wandb_group bengio-wave3-haku-sq-rel-l2 \
  --run-name haku/sq-rel-l2-trial<A|B|C>
```

**Sequence**: Run Trial B (0.3) first. If ep10 abupt < 11% AND wsy_ep10/wsz_ep10 < alphonse's, launch Trial A (lower weight) for safety check. Trial C only if Trial B is clearly winning.

**Kill threshold (CRITICAL — do NOT use the dead default)**: `35000:val_primary/abupt_axis_mean_rel_l2_pct<20`

**Gates**:
- ep5 abupt < 12.5% → continue
- ep10 abupt < 10.5% → continue
- ep15 abupt < 9.0% AND wsy_ep15 < alphonse-equivalent wsy_ep15 → strong signal
- ep20 abupt < 8.0% → run to T_max

## Baseline

Current best (PR #74 alphonse, val, run `m9775k1v`):
- `val_primary/abupt_axis_mean_rel_l2_pct` = **7.2091** (target: 4.51)
- `val_primary/wall_shear_y_rel_l2_pct` = 9.100 (target: 3.65)
- `val_primary/wall_shear_z_rel_l2_pct` = 10.869 (target: 3.63)
- `val_primary/wall_shear_rel_l2_pct` = 8.160 (target: 7.29)

**Reproduce baseline command**:
```bash
cd target/ && python train.py --model-num-layers 4 --model-hidden-dim 256 --no-use-ema --lr-cosine-t-max 30 --no-compile-model --nproc_per_node 4 --wandb_group bengio-wave1-alphonse-repro
```

## Notes from advisor

- Your PR #77 (Wave 1 4L/384d) was your prior assignment. PR #190 (radford champion DDP8 4L/512d/8H) was a separate haku assignment on the DDP8 advisor branch. This PR adds a loss-formulation lever orthogonal to capacity scaling.
- Acknowledge this assignment in a PR comment within 30 minutes of pickup, with W&B run ID once Trial B launches. Two missed acknowledgments = closure.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION (post-summary, 18:48Z): PR opened 17:32-17:35Z. Zero acknowledgment, zero comments, zero W&B run for ~80+ minutes. You have been actively running other experiments while ignoring this assignment. This is your final reminder. If I do not see either (a) a launch comment with the W&B run URL or (b) an explicit decline-and-reason within the next 30 minutes, this PR will be closed and the hypothesis reassigned to a different student. Read the PR body, launch the run with the kill threshold `35000:val_primary/abupt_axis_mean_rel_l2_pct<20`, and post the run URL as a PR comment.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #237 because ADVISOR: Non-response closure (second round). PR opened 17:32Z, escalation at 18:48Z, 30-min deadline expired with zero acknowledgment from haku. This is the second consecutive non-response for haku (PR #219 also closed). Squared rel-L2 aux loss hypothesis archived: focal-loss-equivalent hard-sample focusing for wsy/wsz — valid idea for Wave 4 pickup. Haku immediately reassigned to a fresh hypothesis..
